doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,481 @@
+================================================
+NEP 40 — Legacy Datatype Implementation in NumPy
+================================================
+
+:title: Extensible Datatypes for NumPy
+:Author: Sebastian Berg
+:Author: Add...
+:Status: Draft
+:Type: Informational
+:Created: 2019-07-17
+
+
+Abstract
+--------
+
+In preparation of NumPy enhancement proposals 41, 42, and 43, this
+NEP details the current status of NumPy datatypes as of NumPy 1.19.
+It describes some of the technical aspects and concepts necessary to
+motivate the following proposals.
+
+
+Detailed Description
+--------------------
+
+
+Flexible Datatypes
+^^^^^^^^^^^^^^^^^^
+
+Some datatypes are inherently *flexible*.
+Flexible here means the values representable by the datatype depend on the
+specific instance.
+As an example, ``datetime64[ns]`` can represent partial seconds which
+``datetime64[s]`` cannot and ``"S8"`` can represent longer strings than ``"S4"``.
```

Comment:
Yeah, forgot about `np.flexible` (and ignored the C-API a bit, although it is mixed, e.g. `AdaptFlexible` of course also adepts the "generic" datetime). I guess I need a new word, hmmm.

Reply:
Anyway, its incorrect here, you are right. Since it was supposed to describe the current state more. I think "variable lengths" is not a very important distinction for most code.

Reply:
I think I will try to go with "parametric" for meaning "there is no canonical instance and different dtype instances may not be able to represent the exact same values". (used here: https://github.com/numpy/numpy/issues/12518#issuecomment-446059956) At least for now, of course "byte order" is also a parameter so to say, but...